File: examples/ruby/property-initializers/PropertyInitializers.runar.rb
require 'runar'

class PropertyInitializers < Runar::StatefulSmartContract
  prop :count,     Bigint,  default: 0
  prop :max_count, Bigint,  readonly: true
  prop :active,    Boolean, readonly: true, default: true

  def initialize(max_count)
    super(max_count)
    @max_count = max_count
  end

  runar_public amount: Bigint
  def increment(amount)
    assert @active
    @count = @count + amount
    assert @count <= @max_count
  end

  runar_public
  def reset
    @count = 0
  end
end

File: examples/rust/property-initializers/PropertyInitializers.runar.rs
use runar::prelude::*;

#[runar::contract]
struct PropertyInitializers {
    count: Int,
    #[readonly]
    max_count: Int,
    #[readonly]
    active: Bool,
}

impl PropertyInitializers {
    fn init(&mut self) {
        self.count = 0;
        self.active = true;
    }

    pub fn increment(&mut self, amount: Int) {
        assert!(self.active);
        self.count = self.count + amount;
        assert!(self.count <= self.max_count);
    }

    pub fn reset(&mut self) {
        self.count = 0;
    }
}
